managedtenants/utils/ocm.py

```diff
@@ -117,6 +131,11 @@ def disable_addon(self, addon_id):
         )
 
     def upsert_addon(self, metadata):
+        if metadata.get("addonImageSetVersion"):
+            try:
+                addon = self.add_addon_version(metadata)
+            except OCMAPIError as exception:
+                raise exception
```

Comment:
Do we need to fallback to a _patch request if response code is 409? Similar to the `add_addon` case? It is poorly documented (api.openshift.com) but feels the behavior of OCM should be the same as the non-versioned case.

```python
self._patch(f"/api/clusters_mgmt/v1/addons/{addon_id}/versions/{version_id}", json=imageset)
```

Having some OCM integration tests would be quite useful to validate this. It's probably quite involved, but there is a way to run it locally in a container. WDYT?

Suggested refactor (drop extra try/except):

```python
def upsert_addon(self, metadata):
    try:
        if metadata.get("addonImageSetVersion"):
            addon = self.add_addon_version(metadata)
        else:
            addon = self.add_addon(metadata)
    except OCMAPIError as exception:
        if exception.response.status_code == 409:
            if metadata.get("addonImageSetVersion"):
                return self.update_addon_version(metadata)
            else:
                return self.update_addon(metadata)
        raise exception

    return addon
```

Reply:
> Suggested refactor (drop extra try/except):

@sugarraysam  This refactor would allow the addon only to only be posted to either endpoints. What we have to do instead is:
- Post the addon to both endpoints if it uses ImageSets
- Post the addon only to the old endpoint if it used indexImage field

Reference issue - https://issues.redhat.com/browse/MTSRE-214

Reply:
I just refactored this a bit that reduced duplicate code. Let me know what you think :)

Reply:
:heart: im a big fan of that refactor. I totally mislead you , please never trust me haha